champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/ranger/bow/BioticShot.java

```diff
@@ -0,0 +1,241 @@
+        CustomDamageEvent cde = event.getCustomDamageEvent();
+        if (!(cde.getProjectile() instanceof Arrow arrow)) return;
+        if (!(cde.getDamager() instanceof Player damager)) return;
+        if (!arrows.contains(arrow)) return;
+
+        upwardsArrows.remove(damager);
+
+        int level = getLevel(damager);
+        if (level > 0) {
+            onHit(damager, cde.getDamagee(), level);
+            arrows.remove(arrow);
+            arrow.remove();
+            cde.addReason(getName());
+            if (UtilEntity.isEntityFriendly(damager, cde.getDamagee())) {
+                event.setCancelled(true);
+            }
+        }
+    }
+
+    @EventHandler
+    public void onProjectileLaunch(ProjectileLaunchEvent event) {
+        if (event.getEntity() instanceof Arrow arrow && arrow.getShooter() instanceof Player shooter) {
+            Vector initialVelocity = arrow.getVelocity();
+            int level = getLevel(shooter);
+
+            double totalMagnitude = initialVelocity.length();
+
+            if (level > 0 && initialVelocity.getY() / totalMagnitude >= 0.5) {
+                upwardsArrows.put(shooter, arrow);
+                initialVelocities.put(arrow, initialVelocity);
+            }
+        }
+    }
+
+
+    @UpdateEvent
+    public void checkPlayerHitboxes() {
+        Iterator<Map.Entry<Player, Arrow>> iterator = upwardsArrows.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<Player, Arrow> entry = iterator.next();
+            Player shooter = entry.getKey();
+            Arrow arrow = entry.getValue();
+            if (!arrows.contains(arrow)) continue;
+
+            Vector initialVelocity = initialVelocities.get(arrow);
+
+            if (arrow.getVelocity().getY() < 0 && initialVelocity != null && initialVelocity.length() < 0.5) {
+
+                RayTraceResult result = arrow.getWorld().rayTraceEntities(
+                        arrow.getLocation(),
+                        arrow.getLocation().getDirection(),
+                        0.5,
+                        0.2,
+                        entity -> entity instanceof LivingEntity
+                );
+
+                if (result != null && result.getHitEntity() != null && result.getHitEntity().equals(shooter)) {
+                    Player target = (Player) result.getHitEntity();
+                    int level = getLevel(shooter);
+                    onHit(target, target, level);
+                    iterator.remove();
+                    initialVelocities.remove(arrow);
+                    upwardsArrows.remove(shooter);
+                    arrow.remove();
+                }
+            }
+        }
+    }
+
+    @EventHandler
+    public void onArrowHit(ProjectileHitEvent event) {
+        if (!(event.getEntity() instanceof Arrow arrow)) return;
+        if (!(arrow.getShooter() instanceof Player player)) return;
+        if (!hasSkill(player)) return;
+        if (!upwardsArrows.containsValue(arrow)) return;
```

Comment:
This is a slow check, it is unironically faster to do

Arrow upwardsArrow = upwardsArrows.get(player);
if(upwardsArrow == null) return;
if(!upwardsArrow.equals(arrow)) return;

upwardsArrows.remove(player);